### PR TITLE
fix: [go_router] Extra parameter is always null in route level redirect callback

### PR DESCRIPTION
**Pull request by Semapl3**
fixes https://github.com/flutter/flutter/issues/106164

### CODE REVIEW

**Comment by DamienMrtl** [comment]
Please allow this merge it fixes a regression

**Review by chunhtai — COMMENTED** [review]
Hi @Semapl3 Thanks for the fix, can you add a test to make sure we don't accidentally revert it?

File: packages/go_router/CHANGELOG.md [hunk 1]
```diff
@@ -1,3 +1,7 @@
+## 4.2.2
+
+- Add test extra parameter
```
Comment (chunhtai) [inline packages/go_router/CHANGELOG.md:3]: Can you add more detail on what this pr does?

**Comment by Semapl3** [comment]
@chunhtai, fixed for 4.2.1 and test ready. Check please

File: packages/go_router/test/go_router_test.dart [hunk 2]
```diff
@@ -1170,6 +1170,46 @@ void main() {
...
+          builder: (BuildContext context, GoRouterState state) =>
+              const HomeScreen(),
+          routes: <GoRoute>[
+            GoRoute(
+              name: 'login',
+              path: 'login',
+              builder: (BuildContext context, GoRouterState state) {
+                return const LoginScreen();
+              },
+              redirect: (GoRouterState state) {
```
Comment (chunhtai) [inline packages/go_router/test/go_router_test.dart:1188]: can you add a check to make sure this redirect is called?

File: packages/go_router/test/go_router_test.dart [hunk 3]
```diff
@@ -1170,6 +1170,46 @@ void main() {
...
+            ),
+          ],
+        ),
+      ];
+
+      final GoRouter router = await createRouter(
+        routes,
+        tester,
+        redirect: (GoRouterState state) {
+          if (state.location == '/login') {
```
Comment (chunhtai) [inline packages/go_router/test/go_router_test.dart:1202]: same here

**Comment by Calamity34** [comment]
bump

**Review by chunhtai — APPROVED** [review]
LGTM, it looks like there is a merge conflict, please rebase off the latest main branch

**Comment by chunhtai** [comment]
cc john for secondary review

**Review by johnpryan — APPROVED** [review]

**Comment by johnpryan** [comment]
It looks like this is breaking tests in go_router_builder

**Comment by Semapl3** [comment]
https://github.com/flutter/packages/pull/2404

**Pull request closed without merging**